core/blockchain.go

```diff
@@ -1050,12 +1105,12 @@ func (self *BlockChain) InsertChain(chain types.Blocks) (int, error) {
 
 		if BadHashes[block.Hash()] {
 			err := BadHashError(block.Hash())
-			blockErr(block, err)
+			reportBlock(block, err)
 			return i, err
 		}
-		// Call in to the block processor and check for errors. It's likely that if one block fails
-		// all others will fail too (unless a known block is returned).
-		logs, receipts, err := self.processor.Process(block)
+		// Stage 1 validation of the block using the chain's validator
+		// interface.
+		err := self.validator.ValidateBlock(block)
```

Comment:
Similarly to header insertion, these validator accesses should also be protected by a sync mechanism. I'd suggest moving the validator/processor retrieval to the beginning of the function and wrap it around a custom rw mutex.